In [1]:
import pandas as pd

In [2]:
train = pd.read_csv("train.csv")

In [3]:
train

bidder_id  \
0     91a3c57b13234af24875c56fb7e2b2f4rb56a   
1     624f258b49e77713fc34034560f93fb3hu3jo   
2     1c5f4fc669099bfbfac515cd26997bd12ruaj   
3     4bee9aba2abda51bf43d639013d6efe12iycd   
4     4ab12bc61c82ddd9c2d65e60555808acqgos1   
...                                     ...   
2008  369515b3af4f8ca582f90271d30b14b6r52aw   
2009  f939c17ffc7c39ac9b35b69e5e75179fv9pe2   
2010  c806dbb2decba0ed3c4ff5e2e60a74c2wjvbl   
2011  0381a69b7a061e9ace2798fd48f1f537mgq57   
2012  84a769adc98498f52debfe57b93a0789556f4   

                            payment_account  \
0     a3d2de7675556553a5f08e4c88d2c228754av   
1     a3d2de7675556553a5f08e4c88d2c228v1sga   
2     a3d2de7675556553a5f08e4c88d2c2280cybl   
3     51d80e233f7b6a7dfdee484a3c120f3b2ita8   
4     a3d2de7675556553a5f08e4c88d2c22857ddh   
...                                     ...   
2008  a1f85275793c4a782f0a668711f41b927ivc9   
2009  a3d2de7675556553a5f08e4c88d2c2286s1m2   
2010  a3d2de7675556553a5f08e4c88d2c22856leq   
2011  fd87037ce0304077079c749f420f0b4c54uo0   
2012  fbe0ce34d6546ebd9e4c63afc68b085byd2tf   

                                    address  outcome  
0     a3d2de7675556553a5f08e4c88d2c228vt0u4      0.0  
1     ae87054e5a97a8f840a3991d12611fdcrfbq3      0.0  
2     92520288b50f03907041887884ba49c0cl0pd      0.0  
3     4cb9717c8ad7e88a9a284989dd79b98dbevyi      0.0  
4     2a96c3ce94b3be921e0296097b88b56a7x1ji      0.0  
...                                     ...      ...  
2008  e6882cf204a9482edd042b6e31791dfctxzx8      0.0  
2009  b9b03d5a127eb07aeb9163cdcf524e1344ac9      0.0  
2010  d02c2b288b8aabd79ff47118aff41a2dqwzwc      0.0  
2011  f030a221726fbcdfc4dc7dfd1b381a112hieq      0.0  
2012  a3d2de7675556553a5f08e4c88d2c228fib6p      0.0  

[2013 rows x 4 columns]

In [4]:
train["outcome"].unique()

array([0., 1.])

In [5]:
train.dtypes

bidder_id           object
payment_account     object
address             object
outcome            float64
dtype: object

**Définition des variables :**
- Bidder_id : l’id d’echérisseur.
- Payement_account : Le compte qui a effectué la transaction.
- Adresse : l’adresse de d’echérisseur.
- Outcome : 0 si une personne, 1 si robot.

In [6]:
train.nunique()

bidder_id          2013
payment_account    2013
address            2013
outcome               2
dtype: int64

+ Il apparait bien que tout les bidder_id du dataset train sont unique.

In [7]:
bids=pd.read_csv("bids.csv")

In [8]:
bids

bid_id                              bidder_id auction  \
0              0  8dac2b259fd1c6d1120e519fb1ac14fbqvax8   ewmzr   
1              1  668d393e858e8126275433046bbd35c6tywop   aeqok   
2              2  aa5f360084278b35d746fa6af3a7a1a5ra3xe   wa00e   
3              3  3939ac3ef7d472a59a9c5f893dd3e39fh9ofi   jefix   
4              4  8393c48eaf4b8fa96886edc7cf27b372dsibi   jefix   
...          ...                                    ...     ...   
7656329  7656329  626159dd6f2228ede002d9f9340f75b7puk8d   3e64w   
7656330  7656330  a318ea333ceee1ba39a494476386136a826dv   xn0y0   
7656331  7656331  f5b2bbad20d1d7ded3ed960393bec0f40u6hn   gja6c   
7656332  7656332  d4bd412590f5106b9d887a43c51b254eldo4f   hmwk8   
7656333  7656333  0ea62aaa9c3ffcc6db584cb69c1f6c4bcripp   c9ox9   

            merchandise    device              time country              ip  \
0               jewelry    phone0  9759243157894736      us   69.166.231.58   
1             furniture    phone1  9759243157894736      in   50.201.125.84   
2            home goods    phone2  9759243157894736      py  112.54.208.157   
3               jewelry    phone4  9759243157894736      in   18.99.175.133   
4               jewelry    phone5  9759243157894736      in    145.138.5.37   
...                 ...       ...               ...     ...             ...   
7656329         jewelry   phone91  9709222052631578      ru  140.204.227.63   
7656330          mobile  phone236  9709222052631578      pl  24.232.159.118   
7656331  sporting goods   phone80  9709222052631578      za   80.237.28.246   
7656332         jewelry  phone349  9709222052631578      my   91.162.27.152   
7656333          mobile   phone82  9709222052631578      jo  160.243.101.60   

                     url  
0        vasstdc27m7nks3  
1        jmqlhflrzwuay9c  
2        vasstdc27m7nks3  
3        vasstdc27m7nks3  
4        vasstdc27m7nks3  
...                  ...  
7656329  cghhmomsaxi6pug  
7656330  wgggpdg2gx5pesn  
7656331  5xgysg14grlersa  
7656332  bhtrek44bzi2wfl  
7656333  vasstdc27m7nks3  

[7656334 rows x 9 columns]

In [9]:
bids.dtypes

bid_id          int64
bidder_id      object
auction        object
merchandise    object
device         object
time            int64
country        object
ip             object
url            object
dtype: object

**Définition des variables :**
- Bid_id : id de la transaction.
- bidder_id : id d’echérisseur.
- auction : nom du produit acheté.
- merchandise : catégorie d’achat.
- device : Appareil de la quel est effectué la transaction.
- time : temps de la transaction.
- country : le pays de la transaction.
- ip : adresse ip.
- url : --.

In [10]:
bids.nunique()

bid_id         7656334
bidder_id         6614
auction          15051
merchandise         10
device            7351
time            776529
country            199
ip             2303991
url            1786351
dtype: int64

+ Il y a des bidder_id qui se répètent, or il y a que 6614 qui sont unique  

+ Prélévement (filtrage) des enchères dans le dataset "bids" dont l'attribut 'bidder_id' est présent dans le dataset "train" :

In [11]:
unique_bidder_ids = train["bidder_id"].unique()
len(unique_bidder_ids)

2013

In [12]:
filtred_bids= bids[bids["bidder_id"].isin(unique_bidder_ids)]
len(filtred_bids)

3071224

- Donc,il y a 3071224 bidder_id dans bids qui font partie de train. 

In [13]:
filtred_bids

bid_id                              bidder_id auction  \
1              1  668d393e858e8126275433046bbd35c6tywop   aeqok   
5              5  e8291466de91b0eb4e1515143c7f74dexy2yr   3vi4t   
9              9  a58ace8b671a7531c88814bc86b2a34cf0crb   3zpkj   
10            10  a58ace8b671a7531c88814bc86b2a34cf0crb   z6j8b   
13            13  7e34a26c4c339e875997f4b9239d2b5fscng4   nbx39   
...          ...                                    ...     ...   
7656321  7656321  f5b2bbad20d1d7ded3ed960393bec0f40u6hn   gja6c   
7656323  7656323  9655ccc7c0c193f1549475f02c54dce45kjw7   1f9t0   
7656325  7656325  93e3c34171ab65e0b031dc67461e48b478u36   jmny0   
7656326  7656326  03aafab9868455b78f0723eda86698ba9v9rm   lw44y   
7656331  7656331  f5b2bbad20d1d7ded3ed960393bec0f40u6hn   gja6c   

            merchandise    device              time country              ip  \
1             furniture    phone1  9759243157894736      in   50.201.125.84   
5                mobile    phone7  9759243157894736      ru   91.107.221.27   
9        sporting goods    phone4  9759243210526315      za  123.28.123.226   
10       sporting goods   phone10  9759243210526315      ng   209.80.85.101   
13               mobile   phone13  9759243263157894      in  143.183.225.33   
...                 ...       ...               ...     ...             ...   
7656321  sporting goods    phone1  9709222000000000      za  231.197.103.67   
7656323  sporting goods  phone718  9709222000000000      us   199.76.204.89   
7656325         jewelry  phone892  9709222052631578      id  136.144.47.247   
7656326          mobile   phone16  9709222052631578      in   5.146.235.202   
7656331  sporting goods   phone80  9709222052631578      za   80.237.28.246   

                     url  
1        jmqlhflrzwuay9c  
5        vasstdc27m7nks3  
9        vasstdc27m7nks3  
10       adgjz71s3p2oiau  
13       e5i17pk9wygxewv  
...                  ...  
7656321  pqzlexr3qrl6gmx  
7656323  vasstdc27m7nks3  
7656325  36phop3q8vduhf5  
7656326  oqknoadbcbcm6so  
7656331  5xgysg14grlersa  

[3071224 rows x 9 columns]

In [14]:
filtred_bids.isna().sum()

bid_id            0
bidder_id         0
auction           0
merchandise       0
device            0
time              0
country        2701
ip                0
url               0
dtype: int64

**Extraction des informations pertinentes pour l'entrainement de nos modèles :**

D'après l'analyse de la nouvelle filtred_bids il se voit qu'on peut extraire des variables qui semble importantes lors du developpement du modèle: 
- Une variable **total_bids** pour stocker le nombre de transactions totale effectué par le même bidder. Cette caractéristique capture l'activité globale d'un bidder.
- Une variable **avg_time_between_bids** la moyenne du temps entre les transactions. Cette caractéristique peut indiquer le comportement de soumission des enchères d'un bidder, s'il est actif de manière régulière ou sporadique.
- Une variable **most_frequent_merchandise** la catégorie fréquemment choisit par le bidder. Cela peut révéler les préférences d'un bidder en termes de types d'objets auxquels il participe.
- Une variable **most_frequent_country** le pays le plus fréquent.
- Une variable **unique_auctions** Le nombre d'enchères uniques auxquelles un bidder a participé. Cela donne un aperçu de la diversité des enchères auxquelles un bidder participe.
- Une variable **unique_device_ratio** le ratio du nombre de devices uniques utilisés par un bidder par rapport au nombre total de bids. Cela peut donner des informations sur la variété des devices utilisés par un bidder.


In [15]:
# Calcul du nombre de transactions total efféctué par un meme bidder
total_bids= filtred_bids.groupby('bidder_id')['bid_id'].count().reset_index()
total_bids.columns =['bidder_id','total_bids']

# Calcul du temps moyens entre les différentes transaction
avg_time_between_bids=filtred_bids.groupby('bidder_id')['time'].diff().groupby(filtred_bids['bidder_id']).mean().reset_index()
avg_time_between_bids.columns =['bidder_id','avg_time_between_bids']

# Calcul de la catégorie fréquement choisit
most_frequent_merchandise= filtred_bids.groupby('bidder_id')['merchandise'].agg(lambda x: x.value_counts().index[0] if not 
                                                                                x.empty else None).reset_index()
most_frequent_merchandise.columns =['bidder_id','most_frequent_merchandise']

# Calcul du Pays frequement utilisé par le bidder
most_frequent_country= filtred_bids.groupby('bidder_id')['country'].agg(lambda x: x.value_counts().index[0] if not x.empty 
                                                                        and len(x.value_counts()) > 0 else None).reset_index()
most_frequent_country.columns =['bidder_id','most_frequent_country']

# Calcul du nombre d'enchère unique
unique_auctions=filtred_bids.groupby('bidder_id')['auction'].nunique().reset_index()
unique_auctions.columns =['bidder_id','unique_auctions']

#Calcul du ratio des devices unique par rapport au total devices utilisé

unique_device_ratio= filtred_bids.groupby('bidder_id')['device'].nunique()/filtred_bids.groupby('bidder_id')['bid_id'].count()
unique_device_ratio=unique_device_ratio.reset_index()
unique_device_ratio.columns = ['bidder_id', 'unique_device_ratio']



- Maintenant, c'est le temps de construire la nouvelle trainset en combinant les différent variables

In [16]:
unique_device_ratio

bidder_id  unique_device_ratio
0     001068c415025a009fee375a12cff4fcnht8y             1.000000
1     0030a2dd87ad2733e0873062e4f83954mkj86             1.000000
2     00a0517965f18610417ee784a05f494d4dw6e             0.475177
3     00cc97158e6f4cb8eac3c0075918b7ffi5k8o             1.000000
4     01067975436d123f717ee5aba0dd4bbfa0937             0.303867
...                                     ...                  ...
1979  ff92ea4abd33ed38601287f0e1d6726dmgx1f             0.250633
1980  ffacbed056cbfaa60c1fcf51f0d381bddr3ly             0.777778
1981  ffaf0a972a6dcb3910fd6b16045781e2ava5y             0.666667
1982  ffc4e2dd2cc08249f299cab46ecbfacfobmr3             0.590909
1983  fff2c070d8200e0a09150bd81452ce29ngcnv             1.000000

[1984 rows x 2 columns]

In [17]:
#Fusion des différentes variables en un Dataframe unique
new_features=pd.merge(total_bids,unique_auctions,on='bidder_id',how='left')
new_features=pd.merge(new_features,avg_time_between_bids,on='bidder_id',how='left')
new_features=pd.merge(new_features,most_frequent_merchandise,on='bidder_id',how='left')
new_features=pd.merge(new_features,most_frequent_country,on='bidder_id',how='left')
new_features=pd.merge(new_features,unique_device_ratio,on='bidder_id',how='left')

#Remplaçement des NAN
new_features=new_features.fillna(0)
print(new_features.isna().sum())

bidder_id                    0
total_bids                   0
unique_auctions              0
avg_time_between_bids        0
most_frequent_merchandise    0
most_frequent_country        0
unique_device_ratio          0
dtype: int64


In [18]:
#Affichage du nouveau Dataframe
new_features

bidder_id  total_bids  unique_auctions  \
0     001068c415025a009fee375a12cff4fcnht8y           1                1   
1     0030a2dd87ad2733e0873062e4f83954mkj86           1                1   
2     00a0517965f18610417ee784a05f494d4dw6e         141               74   
3     00cc97158e6f4cb8eac3c0075918b7ffi5k8o           3                3   
4     01067975436d123f717ee5aba0dd4bbfa0937         543               17   
...                                     ...         ...              ...   
1979  ff92ea4abd33ed38601287f0e1d6726dmgx1f         395               11   
1980  ffacbed056cbfaa60c1fcf51f0d381bddr3ly           9                9   
1981  ffaf0a972a6dcb3910fd6b16045781e2ava5y           3                3   
1982  ffc4e2dd2cc08249f299cab46ecbfacfobmr3          22               15   
1983  fff2c070d8200e0a09150bd81452ce29ngcnv           1                1   

      avg_time_between_bids most_frequent_merchandise most_frequent_country  \
0              0.000000e+00                   jewelry                    bn   
1              0.000000e+00                    mobile                    ir   
2              5.486831e+11           books and music                    in   
3              2.748400e+13          office equipment                    id   
4              2.507283e+10                   jewelry                    in   
...                     ...                       ...                   ...   
1979           1.949637e+11            sporting goods                    lk   
1980           8.825230e+12            sporting goods                    my   
1981           4.388158e+12            sporting goods                    de   
1982           3.596346e+12                    mobile                    ng   
1983           0.000000e+00                home goods                    th   

      unique_device_ratio  
0                1.000000  
1                1.000000  
2                0.475177  
3                1.000000  
4                0.303867  
...                   ...  
1979             0.250633  
1980             0.777778  
1981             0.666667  
1982             0.590909  
1983             1.000000  

[1984 rows x 7 columns]

- Création du nouveau train set :

In [19]:
train = pd.merge(train,new_features,on='bidder_id',how='left')
train = train.fillna(0)

In [20]:
train

bidder_id  \
0     91a3c57b13234af24875c56fb7e2b2f4rb56a   
1     624f258b49e77713fc34034560f93fb3hu3jo   
2     1c5f4fc669099bfbfac515cd26997bd12ruaj   
3     4bee9aba2abda51bf43d639013d6efe12iycd   
4     4ab12bc61c82ddd9c2d65e60555808acqgos1   
...                                     ...   
2008  369515b3af4f8ca582f90271d30b14b6r52aw   
2009  f939c17ffc7c39ac9b35b69e5e75179fv9pe2   
2010  c806dbb2decba0ed3c4ff5e2e60a74c2wjvbl   
2011  0381a69b7a061e9ace2798fd48f1f537mgq57   
2012  84a769adc98498f52debfe57b93a0789556f4   

                            payment_account  \
0     a3d2de7675556553a5f08e4c88d2c228754av   
1     a3d2de7675556553a5f08e4c88d2c228v1sga   
2     a3d2de7675556553a5f08e4c88d2c2280cybl   
3     51d80e233f7b6a7dfdee484a3c120f3b2ita8   
4     a3d2de7675556553a5f08e4c88d2c22857ddh   
...                                     ...   
2008  a1f85275793c4a782f0a668711f41b927ivc9   
2009  a3d2de7675556553a5f08e4c88d2c2286s1m2   
2010  a3d2de7675556553a5f08e4c88d2c22856leq   
2011  fd87037ce0304077079c749f420f0b4c54uo0   
2012  fbe0ce34d6546ebd9e4c63afc68b085byd2tf   

                                    address  outcome  total_bids  \
0     a3d2de7675556553a5f08e4c88d2c228vt0u4      0.0        24.0   
1     ae87054e5a97a8f840a3991d12611fdcrfbq3      0.0         3.0   
2     92520288b50f03907041887884ba49c0cl0pd      0.0         4.0   
3     4cb9717c8ad7e88a9a284989dd79b98dbevyi      0.0         1.0   
4     2a96c3ce94b3be921e0296097b88b56a7x1ji      0.0       155.0   
...                                     ...      ...         ...   
2008  e6882cf204a9482edd042b6e31791dfctxzx8      0.0        36.0   
2009  b9b03d5a127eb07aeb9163cdcf524e1344ac9      0.0         1.0   
2010  d02c2b288b8aabd79ff47118aff41a2dqwzwc      0.0         2.0   
2011  f030a221726fbcdfc4dc7dfd1b381a112hieq      0.0         1.0   
2012  a3d2de7675556553a5f08e4c88d2c228fib6p      0.0         2.0   

      unique_auctions  avg_time_between_bids most_frequent_merchandise  \
0                18.0           5.711121e+11                home goods   
1                 1.0           3.233579e+12          office equipment   
2                 4.0           2.379000e+12            sporting goods   
3                 1.0           0.000000e+00           books and music   
4                23.0           7.777888e+10          office equipment   
...               ...                    ...                       ...   
2008             25.0           2.099162e+12                    mobile   
2009              1.0           0.000000e+00                    mobile   
2010              1.0           9.798474e+12                    mobile   
2011              1.0           0.000000e+00          office equipment   
2012              1.0           1.789474e+09                   jewelry   

     most_frequent_country  unique_device_ratio  
0                       ke             0.583333  
1                       id             0.666667  
2                       id             0.500000  
3                       my             1.000000  
4                       in             0.341935  
...                    ...                  ...  
2008                    uk             0.111111  
2009                    ua             1.000000  
2010                    id             1.000000  
2011                    sa             1.000000  
2012                    ru             0.500000  

[2013 rows x 10 columns]

- Génération du nouveau fichier excel newTrain.csv :

In [23]:
train.to_excel(r'C:\Users\Lenovo\OneDrive\Desktop\SDSI\S2\Data mining\Projet1\new_train.xlsx',index=False)

In [22]:
train['outcome'].value_counts()

outcome
0.0    1910
1.0     103
Name: count, dtype: int64

**Conclusion**
- Voilà par la suite le résultat de notre Feature Extraction, on a parti d'un dataset de 4 variables presentant moins d'informations vers un dataset 'new_train' qui renvoie bcp mieux des statistiques qui pourra nous aider à faire les analyses par la suite.